File: src/BetterLyrics.Sdk/Interfaces/Plugins/IPlugin.cs
using BetterLyrics.Sdk.Models.SettingsSchema;

namespace BetterLyrics.Sdk.Interfaces.Plugins;

public interface IPlugin : IAsyncDisposable
{
    string Title { get; }
    string Description { get; }
    string Author { get; }

    string Id { get; }
    string Version { get; }
    DateTime LastUpdated { get; }

    string RepositoryUrl { get; }

    Task InitializeAsync(IPluginContext context);
    Dictionary<string, SettingDef> GetSettingDefDict();
}

File: src/BetterLyrics.KMP/shared/src/commonMain/kotlin/com/jayfunc/betterlyrics/sdk/interfaces/plugins/IPlugin.kt
package com.jayfunc.betterlyrics.sdk.interfaces.plugins

import com.jayfunc.betterlyrics.sdk.models.settingsschema.SettingDef
import kotlinx.datetime.LocalDateTime

interface IPlugin {
    val title: String
    val description: String
    val author: String

    val id: String
    val version: String
    val lastUpdated: LocalDateTime

    val repositoryUrl: String

    suspend fun initializeAsync(context: IPluginContext)

    fun getSettingDefDict(): Map<String, SettingDef>

    // Equivalent to C#'s IAsyncDisposable.DisposeAsync()
    suspend fun disposeAsync()
}